File: fxgl-eventbus/src/main/java/com/almasb/fxeventbus/Subscriber.java
package com.almasb.fxeventbus;

import javafx.event.Event;
import javafx.event.EventHandler;
import javafx.event.EventType;

/**
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public final class Subscriber {

    private EventBus bus;

    private EventType<? extends Event> eventType;
    private EventHandler<? super Event> eventHandler;

    Subscriber(EventBus bus, EventType<? extends Event> eventType, EventHandler<? super Event> eventHandler) {
        this.bus = bus;
        this.eventType = eventType;
        this.eventHandler = eventHandler;
    }

    /**
     * Stop listening for events.
     */
    public void unsubscribe() {
        bus.removeEventHandler(eventType, eventHandler);
    }
}

File: fxgl-events/src/main/kotlin/com/almasb/fxgl/event/Subscriber.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.event

import javafx.event.Event
import javafx.event.EventHandler
import javafx.event.EventType

/**
 * A subscriber handle for a specific event handler.
 * This handle can be used to unsubscribe (remove) the event handler.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
class Subscriber
internal constructor(private val bus: EventBus,
                     private val eventType: EventType<out Event>,
                     private val eventHandler: EventHandler<in Event>) {

    /**
     * Stop listening for events.
     */
    fun unsubscribe() {
        bus.removeEventHandler(eventType, eventHandler)
    }
}